data_scrapers/sonoma_county.py

```diff
@@ -0,0 +1,216 @@
+#!/usr/bin/env python3
+import requests
+import json
+import re
+from datetime import datetime
+from typing import List, Dict, Union
+from bs4 import BeautifulSoup, element # type: ignore
+
+def get_rows(tag: element.Tag) -> List[element.ResultSet]:
+    """
+    Gets all tr elements in a tag but the first, which is the header
+    """
+    return tag.findAll('tr')[1:]
+
+def get_cells(row: element.ResultSet) -> List[str]:
+    """
+    Gets all th and tr elements within a single tr element
+    """
+    return [el.text for el in row.findAll(['th', 'td'])]
+
+def generate_update_time(soup: BeautifulSoup) -> str:
+    """
+    Generates a timestamp string (e.g. May 6, 2020 10:00 AM) for when the scraper is run
+    """
+    update_time_text = soup.find('time').text.strip()
+    update_datetime = datetime.strptime(update_time_text, '%B %d, %Y %I:%M %p')
+    return update_datetime.isoformat()
+
+def get_source_meta(soup: BeautifulSoup) -> str:
+    """
+    Finds the 'Definitions' header on the page and gets all of the text in it
+    """
+    definitions_header = soup.find('h3', string='Definitions')
+    definitions_text = definitions_header.find_parent().text
+    return definitions_text.replace('\n', ' ')
+
+# apologies for this horror of a output type
+def transform_cases(cases_tag: element.Tag) -> Dict[str, List[Dict[str, Union[str, int]]]]:
+    """
+    Takes in a BeautifulSoup tag for the cases table and returns all cases
+    (historic and active), deaths, and recoveries in the form:
+    { 'cases': [], 'deaths': [], 'recovered': [], 'active': [] }
+    Where each list contains dictionaries (representing each day's data)
+    of form (example for cases):
+    { 'date': '', 'cases': -1, 'cumul_cases': -1 }
+    """
+    cases = []
+    cumul_cases = 0
+    deaths = []
+    cumul_deaths = 0
+    recovered = []
+    cumul_recovered = 0
+    active = []
+    cumul_active = 0
+    rows = get_rows(cases_tag)
+    for row in rows:
+        row_cells = row.findAll(['th', 'td'])
+        # print(type(row_cells))
+        date = row_cells[0].text.replace('/', '-')
+
+        # instead of 0, this dashboard reports the string '-'
+        active_cases, new_infected, dead, recoveries = [0 if el.text == '–' else int(el.text) for el in row_cells[1:]]
+
+        cumul_cases += new_infected
+        cases.append({ 'date': date, 'cases': new_infected, 'cumul_cases': cumul_cases })
+
+        new_deaths = dead - cumul_deaths
+        deaths.append({ 'date': date, 'deaths': new_deaths, 'cumul_deaths': dead })
+
+        new_recovered = recoveries - cumul_recovered
+        recovered.append({ 'date': date, 'recovered': new_recovered, 'cumul_recovered': recoveries })
+
+        new_active = active_cases - cumul_active
+        active.append({ 'date': date, 'active': new_active, 'cumul_active': active_cases })
+
+    return { 'cases': cases, 'deaths': deaths, 'recovered': recovered, 'active': active }
```

Comment:
Do we want to include "active" and "recovered" time-series even if they're not in the data model? I believe Sonoma is the only county with these.  I'm personally suspicious of the label "recovered" when things like strokes are affecting people long after they test as negative. We're also excluding all the detailed breakdowns Sonoma has over symptoms and underlying conditions, which makes me ask the question: if data isn't in our data model, are there criteria we're applying to decide whether to include or exclude it?

Reply:
I put active and recovered in because they were there and I'd rather err on the side of putting too much data in. If we don't want to use them, then I'll take them out of the output.